src/unix/mmap.c

```diff
@@ -680,26 +670,39 @@ static sysreturn mmap(void *target, u64 size, int prot, int flags, int fd, u64 o
     }
 
     /* Paint into process vmap */
-    struct vmap q;
-    q.flags = vmflags;
-    q.node.r = irange(where, where + len);
-    vmap_paint(h, p->vmaps, &q);
-
     if (flags & MAP_ANONYMOUS) {
         thread_log(current, "   anon target: %s, 0x%lx, len: 0x%lx (given size: 0x%lx)",
                    mapped ? "existing" : "new", where, len, size);
-        zero_mapped_pages(where, len);
+        struct vmap q;
+        q.flags = vmflags;
+        q.node.r = irange(where, where + len);
+        vmap_paint(h, p->vmaps, &q);
+        zero_mapped_pages(where, len); // XXX?
         return where;
     }
 
     file f = resolve_fd(current->p, fd);
     thread_log(current, "  read file at 0x%lx, %s map, blocking...", where, mapped ? "existing" : "new");
+    u64 flen = MIN(pad(f->length, PAGESIZE), len);
+
+    /* Paint file portion */
+    struct vmap q;
+    q.flags = vmflags;
+    q.node.r = irange(where, where + flen);
+    vmap_paint(h, p->vmaps, &q);
+
+    /* And any remaining portion as anonymous (temporary until we have true file demand paging / write-back) */
```

Comment:
Do we need two separate regions here? Seems that we don't really rely on the anonymous flag anywhere other than for sanity checking in the page fault handler, which could just drop that check. We'll need to do so anyways once we put support in for demand paging the file-backed portion itself as well

Reply:
That's true, it doesn't necessarily need to be two different regions, and we can just fault in a zeroed page whenever a fault happens in a valid vmap regardless of flags. In the future, the fault handler would be wired up to a buffer / page cache which can propagate writes (dirty pages) back to disk, etc.